tsl/src/compression/recompress.c

```diff
@@ -0,0 +1,725 @@
+Datum
+tsl_recompress_chunk_segmentwise(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_not_compressed = PG_ARGISNULL(1) ? true : PG_GETARG_BOOL(1);
+
+	ts_feature_flag_check(FEATURE_HYPERTABLE_COMPRESSION);
+	TS_PREVENT_FUNC_IF_READ_ONLY();
+	Chunk *chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+
+	if (!ts_chunk_is_partial(chunk))
+	{
+		int elevel = if_not_compressed ? NOTICE : ERROR;
+		elog(elevel,
+			 "nothing to recompress in chunk %s.%s",
+			 NameStr(chunk->fd.schema_name),
+			 NameStr(chunk->fd.table_name));
+	}
+	else
+	{
+		if (!ts_guc_enable_segmentwise_recompression)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("segmentwise recompression functionality disabled, "
+							"enable it by first setting "
+							"timescaledb.enable_segmentwise_recompression to on")));
+		}
+		uncompressed_chunk_id = recompress_chunk_segmentwise_impl(chunk);
+	}
+
+	PG_RETURN_OID(uncompressed_chunk_id);
+}
+
+Oid
+recompress_chunk_segmentwise_impl(Chunk *uncompressed_chunk)
+{
+	Oid uncompressed_chunk_id = uncompressed_chunk->table_id;
+
+	/*
+	 * only proceed if status in (3, 9, 11)
+	 * 1: compressed
+	 * 2: compressed_unordered
+	 * 4: frozen
+	 * 8: compressed_partial
+	 */
+	if (!ts_chunk_is_compressed(uncompressed_chunk) && ts_chunk_is_partial(uncompressed_chunk))
+		elog(ERROR,
+			 "unexpected chunk status %d in chunk %s.%s",
+			 uncompressed_chunk->fd.status,
+			 NameStr(uncompressed_chunk->fd.schema_name),
+			 NameStr(uncompressed_chunk->fd.table_name));
+
+	/* need it to find the segby cols from the catalog */
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
+	CompressionSettings *settings = ts_compression_settings_get(compressed_chunk->table_id);
+
+	/* new status after recompress should simply be compressed (1)
+	 * It is ok to update this early on in the transaction as it keeps a lock
+	 * on the updated tuple in the CHUNK table potentially preventing other transaction
+	 * from updating it
+	 */
+	if (ts_chunk_clear_status(uncompressed_chunk,
+							  CHUNK_STATUS_COMPRESSED_UNORDERED | CHUNK_STATUS_COMPRESSED_PARTIAL))
+		ereport(DEBUG1,
+				(errmsg("cleared chunk status for recompression: \"%s.%s\"",
+						NameStr(uncompressed_chunk->fd.schema_name),
+						NameStr(uncompressed_chunk->fd.table_name))));
+
+	ereport(DEBUG1,
+			(errmsg("acquiring locks for recompression: \"%s.%s\"",
+					NameStr(uncompressed_chunk->fd.schema_name),
+					NameStr(uncompressed_chunk->fd.table_name))));
+	/* lock both chunks, compressed and uncompressed */
+	/* TODO: Take RowExclusive locks instead of AccessExclusive */
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, ExclusiveLock);
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, ExclusiveLock);
+
+	/*
+	 * Calculate and add the column dimension ranges for the src chunk. This has to
```

Comment:
Might be good to mention that this is for the `enable_chunk_skipping()` functionality, the name "column stats" is very generic so I was unsure.